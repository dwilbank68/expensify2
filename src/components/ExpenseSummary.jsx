import React from 'react';
import numeral from 'numeral';
// import PropTypes from 'prop-types';
import {connect} from 'react-redux';

import expensesTotal from '../selectors/expensesTotal.js';
import selectExpenses from '../selectors/expenses.js';

// import ExpenseSummary from './ExpenseSummary.jsx';
// const ExpenseSummary = (props) => {
export const
ExpenseSummary = ({expensesCount, expensesTotal}) => {

    const renderMessage = () => {
        const expenseWord = expensesCount == 1 ? 'expense' : 'expenses';
        return `Viewing ${expensesCount} ${expenseWord} totalling ${expensesTotal}`
    }

    return (
        <div className="expense-summary">
            <h1>{renderMessage()}</h1>
        </div>
    );
};


const mapStateToProps = state => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters)
    return {
        expensesCount: visibleExpenses.length,
        expensesTotal: numeral(expensesTotal(visibleExpenses)/100).format('$0,0.00')
    }
};
///////////////////////////// mapDispatchToProps //////////////////////////////
//
// Option 1. Skip it - dispatch is on props anyway
// -----------> this.props.dispatch(loadCourses());
//
//
// Option 2. Wrap props manually
//
// function mapDispatchToProps(dispatch) {
//     return {
//         loadCourses: () => {dispatch(loadCourses())},
//         createCourse: (course) => {dispatch(createCourse(course))}
//     };
// }
// function mapDispatchToProps(dispatch) => {(
//     loadCourses: () => dispatch(loadCourses()),
//     createCourse: course => dispatch(createCourse(course))
// });
// }
// -> this.props.loadCourses, this.props.createCourse
//
//
// Option 3. use bindActionCreators (which is just a shortcut method)
//
// function mapDispatchToProps(dispatch) {
//     return bindActionCreators(
//         { nameYouWantOnProps:nameOfImportedAction },
//         dispatch
//     );
// }
//
// function mapDispatchToProps(dispatch) {
//     return {
//         actions: bindActionCreators(actions, dispatch)
//     };
// }
//
// use it like -> this.props.actions.loadCourses();

// ExpenseSummary.defaultProps = {};
// ExpenseSummary.propTypes = {
//     name:        PropTypes.string.isRequired,
//     id:          PropTypes.oneOfType([ PropTypes.number, PropTypes.string ]).isRequired,
//     message:     PropTypes.shape({ title: PropTypes.string, text: PropTypes.string }).isRequired,
//     comments:    PropTypes.arrayOf(React.PropTypes.object),
//     date:        PropTypes.instanceOf(Date)
// };
//
// PropTypes -> array, bool, func, number, object, string, symbol

// export default ExpenseSummary;
export default connect(mapStateToProps)(ExpenseSummary);