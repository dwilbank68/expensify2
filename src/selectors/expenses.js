import moment from 'moment';

export default
(arr, {text, sortBy, startDate, endDate}) => (
    arr
        .filter(e => {
            // if startDate is undefined, or not a number, 'true', so that 'e' passes thru
            // const startDateMatch =  typeof startDate !== 'number'   || e.createdAt >= startDate;
            // const endDateMatch =    typeof endDate !== 'number'     || e.createdAt <= endDate;
            const createdAtMoment = moment(e.createdAt);
            const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
            const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
            const textMatch = e.description.toLowerCase().includes(text.toLowerCase());
            return startDateMatch && endDateMatch && textMatch;
        })
        .sort((a, b) => {
            if (sortBy === 'date') {
                return a.createdAt < b.createdAt ? 1 : -1;
            }
            if (sortBy === 'amount') {
                return a.amount < b.amount ? 1 : -1;
            }
        })
)
