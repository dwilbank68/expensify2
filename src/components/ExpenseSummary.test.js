import React from 'react';
import {shallow} from 'enzyme';

import {ExpenseSummary} from './ExpenseSummary.jsx';
import expenses from '../testing/fixtures/expenses.js';

import '../setupTests.js';

test('should render ExpenseSummary with expenses', () => {
    const wrapper = shallow(<ExpenseSummary expensesCount={2}
                                            expensesTotal={'$664.35'}/>);
    expect(wrapper).toMatchSnapshot();
})

test('should render ExpenseSummary with no expenses', () => {
    const wrapper = shallow(<ExpenseSummary expensesCount={0}
                                            expensesTotal={'$0.00'}/>);
    expect(wrapper).toMatchSnapshot();
})