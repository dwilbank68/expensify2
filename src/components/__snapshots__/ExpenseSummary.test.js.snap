// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary with expenses 1`] = `
<div
  className="expense-summary"
>
  <h1>
    Viewing 2 expenses totalling $664.35
  </h1>
</div>
`;

exports[`should render ExpenseSummary with no expenses 1`] = `
<div
  className="expense-summary"
>
  <h1>
    Viewing 0 expenses totalling $0.00
  </h1>
</div>
`;
